### PR TITLE
Add heterogeneous processing of entities 

### PR DESCRIPTION
**Pull request by juliababkina**
Solution for the issue #85

### CODE REVIEW

File: sim/pkg/simulator/through_stock.go [hunk 1]
```diff
@@ -36,10 +35,22 @@ func (s *stock) Count() uint64 {
...
 }
 
-func (s *stock) EntitiesInStock() []*Entity {
+func (s *stock) EntitiesInStock() map[Entity]bool {
 	return s.stock
 }
 
+func (s *stock) GetEntityByNumber(number int) Entity {
+	counter := 0
+	fmt.Println("new loop")
```
Comment (josephburnett) [inline sim/pkg/simulator/through_stock.go:44]: Is this debugging code?

File: sim/pkg/model/replicas_active.go [hunk 2]
```diff
@@ -40,31 +41,52 @@ func (ras *replicasActiveStock) Count() uint64 {
...
 
 func (ras *replicasActiveStock) Add(entity simulator.Entity) error {
-	replica := entity.(Replica)
+	replica := entity.(ReplicaEntity)
 	replica.Activate()
+	ras.replicas = append(ras.replicas, &replica)
 	return ras.delegate.Add(entity)
 }
 
+func (ras *replicasActiveStock) deleteReplica(replica *ReplicaEntity) {
```
Comment (josephburnett) [inline sim/pkg/model/replicas_active.go:73]: The `EntitiesInStock` method returns an entity set (`map[simulator.Entity]bool`).  But internally you store it as a slice.  Which makes it a little awkward because here you have to iterate through the whole list instead of just looking up the entity.

I would almost rather just continue returning a slice of `*simulatorEntity`.  Is the performance improvement of using a map instead of a slice enough to warrant the additional complexity (conversion, confusion, etc..)?  E.g. why does the map return `simulator.Entity` literal instead of a pointer, while other methods take a `*simulator.Entity`?

File: sim/pkg/simulator/stock_types.go [hunk 3]
```diff
@@ -21,11 +21,12 @@ type baseStock interface {
 	Name() StockName
 	KindStocked() EntityKind
 	Count() uint64
-	EntitiesInStock() []*Entity
+	EntitiesInStock() map[Entity]bool
+	GetEntityByNumber(number int) Entity
```
Comment (josephburnett) [inline sim/pkg/simulator/stock_types.go:25]: Who uses this?  Why did you need to add it?

File: sim/pkg/simulator/through_stock.go (hunk 1 above)
Comment (juliababkina) [reply sim/pkg/simulator/through_stock.go]: Yeah, I was just about to do a commit to remove it. Forgot to remove it.

Comment (juliababkina) [reply sim/pkg/simulator/through_stock.go]: Removed it. See commit ab8fea3c6d8c8f9ae7d04e1406121f9279bec19d.

File: sim/pkg/model/replicas_active.go (hunk 2 above)
Comment (josephburnett) [reply sim/pkg/model/replicas_active.go:73]: Chatted with Julia.  We agreed that there should be 2 underlying implementations of the stock, one optimized for homogeneous entities and one optimized for heterogeneous entities.  The difference will be the runtime of the `Add` `Delete` and `GetAll` methods.  Additionally, the homogeneous stock base will promise stability for the `GetAll` method (which will continue to return a slice) so the get by number method won't be needed for round-robin routing.

File: sim/pkg/data/run_store_test.go [hunk 4]
```diff
@@ -81,11 +81,11 @@ func testStorer(t *testing.T, describe spec.G, it spec.S) {
 
 			subject = NewRunStore(conn)
 
-			stock1 = simulator.NewThroughStock("stock 1", "test entity")
-			stock2 = simulator.NewThroughStock("stock 2", "test entity")
-			env.AddToSchedule(simulator.NewMovement("stock 1 -> stock 2", startAt.Add(111*time.Second), stock1, stock2))
-			env.AddToSchedule(simulator.NewMovement("stock 1 -> stock 2", startAt.Add(222*time.Second), stock1, stock2))
-			env.AddToSchedule(simulator.NewMovement("Ignored", env.HaltTime().Add(10*time.Second), simulator.NewSourceStock("Source", "Entity"), simulator.NewSinkStock("Sink", "Entity")))
+			stock1 = simulator.NewHomogenousThroughStock("stock 1", "test entity")
```
Comment (josephburnett) [inline sim/pkg/data/run_store_test.go:84]: Spelling: homogen[e]ous

File: sim/pkg/model/replicas_active.go [hunk 5]
```diff
@@ -41,30 +42,47 @@ func (ras *replicasActiveStock) Count() uint64 {
...
+	if removed {
+		ras.replicas = ras.replicas[:len(ras.replicas)-1]
+	}
+}
+
 func NewReplicasActiveStock(env simulator.Environment) ReplicasActiveStock {
 	return &replicasActiveStock{
 		env:      env,
-		delegate: simulator.NewThroughStock("ReplicasActive", "Replica"),
+		delegate: simulator.NewHeterogenousThroughStock("ReplicasActive", "Replica"),
```
Comment (josephburnett) [inline sim/pkg/model/replicas_active.go:85]: So you are creating a heterogeneous delegate stock, which should have constant time add and delete.  However you are still maintaining a slice of replicas.  Which means you still get O(n) time deleting (see the for loop above).

Why do you maintain two stocks for replicas (delegate and replicas)?

File: sim/pkg/model/replicas_desired_test.go [hunk 6]
```diff
@@ -41,9 +41,9 @@ func testReplicasDesired(t *testing.T, describe spec.G, it spec.S) {
 	var envFake *FakeEnvironment
 
 	it.Before(func() {
-		replicasLaunching = simulator.NewThroughStock("ReplicasLaunching", "Replica")
-		replicasActive = simulator.NewThroughStock("ReplicasActive", "Replica")
-		replicasTerminated = simulator.NewThroughStock("ReplicasTerminated", "Replica")
+		replicasLaunching = simulator.NewHomogenousThroughStock("ReplicasLaunching", "Replica")
+		replicasActive = simulator.NewHeterogenousThroughStock("ReplicasActive", "Replica")
```
Comment (josephburnett) [inline sim/pkg/model/replicas_desired_test.go:45]: Spelling: heterogen[e]ous

File: sim/pkg/simulator/heterogenous_stock.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+func (hs *heterogenousStock) EntitiesInStock() []*Entity {
+	entities := make([]*Entity, 0)
+
+	for entity := range hs.stock {
+		entities = append(entities, &entity)
+	}
+	return entities
+}
+
+func (hs *heterogenousStock) Add(entity Entity) error {
```
Comment (josephburnett) [inline sim/pkg/simulator/heterogenous_stock.go:32]: You might want to add a comment about the runtime complexity of Add / Remove O(1) versus EntitiesInStock O(n).  After all, that's the point of having two different base stocks.

File: sim/pkg/simulator/homogenous_stock.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+			hs.KindStocked(),
+			entity.Kind(),
+		)
+	}
+
+	hs.stock = append(hs.stock, &entity)
+	return nil
+}
+
+func (hs *homogenousStock) Remove(entity *Entity) Entity {
```
Comment (josephburnett) [inline sim/pkg/simulator/homogenous_stock.go:46]: Here you could panic if you are given a non-nil entity, since homogeneous stocks do not support removing a specific entity.

File: sim/pkg/data/run_store_test.go (hunk 4 above)
Comment (juliababkina) [reply sim/pkg/data/run_store_test.go]: Now we have arrayStock and mapStock

File: sim/pkg/model/replicas_active.go (hunk 5 above)
Comment (juliababkina) [reply sim/pkg/model/replicas_active.go]: Deleted a slice of replicas

File: sim/pkg/model/replicas_desired_test.go (hunk 6 above)
Comment (juliababkina) [reply sim/pkg/model/replicas_desired_test.go]: Now we have arrayStock and mapStock

File: sim/pkg/simulator/heterogenous_stock.go (hunk 7 above)
Comment (juliababkina) [reply sim/pkg/simulator/heterogenous_stock.go]: Added

File: sim/pkg/simulator/homogenous_stock.go (hunk 8 above)
Comment (juliababkina) [reply sim/pkg/simulator/homogenous_stock.go]: Not relevant anymore, as we have arrayStock and mapStock which both can remove a particular entity.

File: sim/pkg/simulator/array_stock.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+func (as *arrayStock) KindStocked() EntityKind {
+	return as.stocksKind
+}
+
+func (as *arrayStock) Count() uint64 {
+	return uint64(len(as.stock))
+}
+
+//time complexity O(len(as.stock))
```
Comment (josephburnett) [inline sim/pkg/simulator/array_stock.go:23]: Isn't this O(1)?  Because you're just returning the stock slice as-is.

Comment (juliababkina) [reply sim/pkg/simulator/array_stock.go]: Sure! It's O(1). copy-paste error
